Update README.md

curran opened the pull request:
First pass docs update

curran commented:
First pass of README update done largely with AI, passing the old README and changelog.

@JesusTheHun Would you be able to please review this and flag any major issues? Thanks!

JesusTheHun commented on README.md:
````diff
@@ -23,17 +23,17 @@ This library is distributed only via [NPM](npmjs.com). Install by running
 
 ```javascript
-const { Graph } = require('graph-data-structure');
+const { Graph, serializeGraph, deserializeGraph, topologicalSort, shortestPath } = require('graph-data-structure');
````
Maybe it's type to use ESM ?

curran replied:
```suggestion
import { Graph, serializeGraph, deserializeGraph, topologicalSort, shortestPath } from 'graph-data-structure';
```

JesusTheHun commented on README.md:
```diff
@@ -23,17 +23,17 @@ This library is distributed only via [NPM](npmjs.com). Install by running
 ### ABC
 
-To create a graph instance, invoke **[Graph](#graph)** as a constructor function.
+To create a graph instance, instantiate **[Graph](#graph)** as a class.
```
I would just say :

> Start by creating a new graph object :

curran replied:
```suggestion
Start by creating a new  **[Graph](#graph)** object.
```

JesusTheHun commented on README.md:
````diff
@@ -67,8 +67,8 @@ Here's an example of topological sort with getting dressed (from Cormen et al. "
 ```javascript
-var graph = Graph()
-  .addEdge('socks', 'shoes')
+var graph = new Graph();
````
Move to `const` ? 😅

curran replied:
```suggestion
const graph = new Graph();
```

JesusTheHun commented on README.md:
```diff
@@ -98,29 +98,29 @@ For more detailed example code that shows more methods, have a look at the [test
 Constructs an instance of the graph data structure.
 
-The optional argument _serialized_ is a serialized graph that may have been generated by **[serialize](#serialize)**. If _serialized_ is present, it is deserialized by invoking **[deserialize](#deserialize)**.
+The optional argument _serialized_ is a serialized graph that may have been generated by **[serializeGraph](#serializeGraph)**. If _serialized_ is present, it is deserialized by invoking **[deserializeGraph](#deserializeGraph)**.
```
Nope, to deserialize the graph you have to call `deserializeGraph(mySerializedObject)`. There constructor takes no parameter anymore

curran replied:
```suggestion
The optional argument _serialized_ is a serialized graph that may have been generated by **[serializeGraph](#serializeGraph)**. If _serialized_ is present, it is deserialized by invoking **[deserializeGraph(mySerializedObject)](#deserializeGraph)**.
```

JesusTheHun commented on README.md:
```diff
@@ -98,29 +98,29 @@ For more detailed example code that shows more methods, have a look at the [test
 <a name="add-node" href="#add-node">#</a> <i>graph</i>.<b>addNode</b>(<i>node</i>)
 
-Adds a node to the graph. Returns _graph_ to support method chaining. The argument _node_ is a string identifier that uniquely identifies the node within this graph instance. If a node with the same identifier was already added to the graph, this function does nothing.
+Adds a node to the graph. Returns _graph_ to support method chaining. The argument _node_ can be any object or string that uniquely identifies the node within this graph instance. If a node with the same identifier was already added to the graph, this function does nothing.
```
I would remove the part on the identifier, as we now accept object, it has no meaning anymore. Or maybe "if the node already exists, this function does nothing"

curran replied:
```suggestion
Adds a node to the graph. Returns _graph_ to support method chaining. If the given _node_ was already added to the graph, this function does nothing.
```

JesusTheHun commented on README.md:
```diff
@@ -98,29 +98,29 @@ For more detailed example code that shows more methods, have a look at the [test
 <a name="remove-node" href="#remove-node">#</a> <i>graph</i>.<b>removeNode</b>(<i>node</i>)
 
-Removes the specified node. Returns _graph_ to support method chaining. The argument _node_ is a string identifier for the node to remove. This function also removes all edges connected to the specified node, both incoming and outgoing.
+Removes the specified node. Returns _graph_ to support method chaining. The argument _node_ is a string or object identifier for the node to remove. This function also removes all edges connected to the specified node, both incoming and outgoing.
```
I would also add a warning but node that are object. You have to remove them using the exact same reference as when they were created. One can use `getNode()` to retrieve such reference

curran replied:
```suggestion
Removes the specified node. Returns _graph_ to support method chaining. The argument _node_ is a string or object identifier for the node to remove. This function also removes all edges connected to the specified node, both incoming and outgoing.

**Note:** You have to remove them using the exact same reference as when they were created. One can use getNode() to retrieve such reference.
```

JesusTheHun commented on README.md:
```diff
@@ -132,108 +132,64 @@ Returns `true` if there exists an edge from node _u_ to node _v_. Returns `false
-  - `target` The node identifier string of the target node (**v**).
+  - `source` The node reference of the source node (**u**).
+  - `target` The node reference of the target node (**v**).
   - `weight` The weight of the edge between the source and target nodes.
```
You forgot the edge's properties, if any

JesusTheHun commented on README.md:
```diff
@@ -132,108 +132,64 @@ Returns `true` if there exists an edge from node _u_ to node _v_. Returns `false
-<a name="deserialize" href="#deserialize">#</a> <i>graph</i>.<b>deserialize</b>(<i>serialized</i>)
-
-Deserializes the given serialized graph. Returns _graph_ to support method chaining. The argument _serialized_ is a graph representation with the structure described in **[serialize](#serialize)**. This function iterates over the serialized graph and adds the nodes and links it represents by invoking **[addNode](#add-node)** and **[addEdge](#add-edge)**. The output from **[serialize](#serialize)* ...
+Deserializes the given serialized graph. Returns a new _graph_. The argument _serialized_ is a graph representation with the structure described in **[serializeGraph](#serializeGraph)**.
```
Makes me think that the `serializeGraph` could also use an identity function, to avoid the duplication of the data when using a node object.

curran commented:
OK I went through and addressed most of the feedback. Merging this now for upcoming release.

Please feel free to update the README if you see any issues @JesusTheHun ! Thanks!
